qa/tasks/rsync.py

```diff
@@ -0,0 +1,285 @@
+
+            # rsync data from snapshot, snap is created using written pattern data
+            elif self.snap_enable:
+                # Create file and add data to the file
+                self.my_mnt.write_test_pattern("{}/file_a".format(self.source_dir), self.file_size * 1024 * 1024)
+                
+                snapshot_name = 'snap' + '{}'.format(iteration)
+
+                # Create Snapshoti
+                self.my_mnt.create_snapshot(self.data_dir, snapshot_name)
+                iteration += 1
+
+                snap_shot = self.data_dir + './snap' + snapshot_name
+
+                # Delete snapshot
+                self.my_mnt.run_shell(["rmdir", "{}".format(snap_shot)])
+
+                # Delete the file created in data directory
+                self.my_mnt.run_shell(["rm", "-f", "{}/file_a".format(self.source_dir)])
+
+            # rsync data from workunit IO data
+            elif self.work_unit:
+                try:
+                    self.my_mnt.run_shell(["rsync", "-azvh", "{}".format(self.source_dir), "rsyncdir/dir1/"])
+                except CommandFailedError as e:
+                    if e.exitstatus == 24:
+                        log.info("Some files vanished before they could be transferred")
+                    else:
+                        raise
+                finally:
+                    # Check for event handler stop message
+                    should_stop = self.ctx.workunit_state.observer_should_stop()
+
+            # rsync data from written pattern data
+            else:
+                # Create file and add data to the file
+                self.my_mnt.write_test_pattern("{}/file_a".format(self.source_dir), self.file_size * 1024 * 1024)
+
+                self.my_mnt.run_shell(["rsync", "-azvh", "{}".format(self.source_dir), "rsyncdir/dir{}/".format(iteration)])
+                iteration += 1
+
+                # Delete the file created in data directory
+                self.my_mnt.run_shell(["rm", "-f", "{}/file_a".format(self.source_dir)])
+
+            # Send back stop request to event handler in workunit task.
+            if should_stop:
+                log.debug("I am here")
+                self.my_mnt.run_shell(["rm", "-rf", "rsyncdir/"])
+                self.my_mnt.run_shell(["rm", "-rf", "{}".format(self.source_dir)])
+                self.ctx.workunit_state.stop_observing()
+
+            time.sleep(self.wait_time)
+
+@contextlib.contextmanager
+def task(ctx, config):
+
+    log.info('Beginning rsync...')
+
+    if config is None:
+        config = {}
+    assert isinstance(config, dict), \
+        "rsync task accepts dict for running configuration"
+
+    run_time = config.get('runtime', 0)
+
+    start_rsync = RSync(ctx, config, logger=log.getChild('rsync'))
+    start_rsync.start()
+
+    try:
+        log.debug('Yielding')
+        yield
+        time.sleep(run_time)
+    finally:
+        log.info('joining rsync thread')
+        start_rsync.stop()
```

Comment:
Check for exception here? See MDSThrasher code. But as you are not inheriting  `Thrasher`, you need to define your own exception variable in `RSync`. Also that would allow DaemonWatchdog to monitor `RSync` if needed.

Reply:
@joscollin I have a doubt regarding why we would check, log, set the exception variable and raise the exception here rather than here: https://github.com/ceph/ceph/pull/28472/files/46bc9f2a0be683951919c4f07c8387eb7c76ca95#diff-d0566013b921ead2517f727df305f8c8R131 (which is already done except setting the exception variable) . Wouldn't doing the check in the end after the thread is stopped(start_rsync.stop()) mean that the exception is unhandled( at the time it was captured ) and the threads are allowed to run till completion?